datapusher/jobs.py

```diff
@@ -427,16 +426,24 @@ def push_to_datastore(task_id, input, dry_run=False):
         logger.info('Done.')
         ct = 'application/csv'
 
+    read_exception = None
     try:
         table_set = messytables.any_tableset(tmp, mimetype=ct, extension=ct)
     except messytables.ReadError as e:
-        # try again with format
+        read_exception = e
+
+    # try again with format inferred from url
+    if not table_set.tables:
         tmp.seek(0)
         try:
-            format = resource.get('format')
-            table_set = messytables.any_tableset(tmp, mimetype=format, extension=format)
-        except:
-            raise util.JobError(e)
+            extension = resource.get('format')
+            format = mimetypes.guess_type(url)[0]
+            table_set = messytables.any_tableset(tmp, mimetype=format, extension=extension)
+        except Exception:
+            raise util.JobError(read_exception)
+
+        if not table_set.tables:
+            raise util.JobError("Unable to read any tabular data from the file.")
```

Comment:
This flow should be improved. There are a few issues I see in it right now:
1. There's no need in initilizing variables with None in 429
1. If you catch ReadError in 432 `table_set` might be None resulting in `NameError: name 'tables' is not defined` in 436
2. I suggest refactor to:
```
try:
except messytables.ReadError as read_exception:
       try:
       except Exception:
            raise util.JobError(read_exception)
if not tables_set or table_set.tables:
            raise util.JobError("Unable to read any tabular data from the file.")
 ```

Reply:
1 and 2 are good catches.  I don't think 3 works. The problem is that a ReadError is not thrown if it doesn't find any tables, it fails silently in our case.  So we need to check that the output exists as expected with an if statement.